File: control_cpp/src/ControlApp.cpp
#include "ControlApp.h"

#include <math.h>

#include <sstream>

#include "DataTypes/Matrix44.h"

/**
 * @brief Constructor
 * @param target connection target in the following format: "hostname:port" or "ip:port", e.g. "localhost:5000"
 */
ControlApp::ControlApp(const std::string& target) : AppClient(std::string(APP_NAME), target) {}

/**
 * @brief Gets called on remote app function calls received from the robot control
 * @param function app function data
 */
void ControlApp::AppFunctionHandler(const robotcontrolapp::AppFunction& function) {}

/**
 * @brief Gets called on remote UI update requests received from the robot control
 * @param updates updated UI elements. Key is the element name, value contains the changes.
 */
void ControlApp::UiUpdateHandler(const std::map<std::string, const robotcontrolapp::AppUIElement*>& updates)
{
    for (auto& update : updates)
    {
        // Handle buttons
        if (update.second->state().state_case() == robotcontrolapp::AppUIElement::AppUIState::kButtonState)
        {
            bool isClicked = update.second->state().button_state() == robotcontrolapp::ButtonState::CLICKED;
            if (isClicked)
            {
                // Init
                if (update.first == "buttonReset")
                    ResetErrors();
                else if (update.first == "buttonEnable")
                    EnableMotors();
                else if (update.first == "buttonDisable")
                    DisableMotors();
                else if (update.first == "buttonReferenceAll")
                    ReferenceAllJoints();
                else if (update.first == "buttonReferenceA1")
                    ReferenceRobotJoint(0);
                else if (update.first == "buttonReferenceProgram")
                    ReferenceAllJoints(true);

                // Velocity override
                else if (update.first == "buttonFaster")
                    ExampleFaster();
                else if (update.first == "buttonSlower")
                    ExampleSlower();

                // Programs
                else if (update.first == "buttonProgramStart")
                    StartMotionProgram();
                else if (update.first == "buttonProgramStop")
                    StopMotionProgram();
                else if (update.first == "buttonProgramPause")
                    PauseMotionProgram();
                else if (update.first == "buttonProgramSingle")
                    SetMotionProgramSingle();
                else if (update.first == "buttonProgramRepeat")
                    SetMotionProgramRepeat();
                else if (update.first == "buttonProgramStep")
                    SetMotionProgramStep();
                else if (update.first == "buttonMotionProgramLoad")
                    LoadMotionProgram(m_motionProgramFile);
                else if (update.first == "buttonMotionProgramUnload")
                    UnloadMotionProgram();
                else if (update.first == "buttonLogicProgramLoad")
                    LoadLogicProgram(m_logicProgramFile);
                else if (update.first == "buttonLogicProgramUnload")
                    UnloadLogicProgram();

                // Move To
                else if (update.first == "buttonMoveToStop")
                    MoveToStop();
                else if (update.first == "buttonMoveToJoint")
                    ExampleMoveToJoint();
                else if (update.first == "buttonMoveToJointRelative")
                    ExampleMoveToJointRelative();
                else if (update.first == "buttonMoveToCart")
                    ExampleMoveToCart();
                else if (update.first == "buttonMoveToCartBaseRelative")
                    ExampleMoveToCartRelativeBase();
                else if (update.first == "buttonMoveToCartToolRelative")
                    ExampleMoveToCartRelativeTool();
                else if (update.first == "buttonProgramUploadSampleFile")
                    ExampleUploadSampleProgramFromFile();
                else if (update.first == "buttonProgramUploadSampleMemory")
                    ExampleUploadSampleProgramFromMemory();
                else if (update.first == "buttonProgramDownloadSampleFile")
                    ExampleDownloadSampleProgramToFile();
                else if (update.first == "buttonProgramDownloadSampleMemory")
                    ExampleDownloadSampleProgramToMemory();
                else if (update.first == "buttonProgramList")
                    ExampleListPrograms();

                // Digital IO
                else if (update.first == "buttonDIn22True")
                    SetDigitalInput(21, true);
                else if (update.first == "buttonDIn22False")
                    SetDigitalInput(21, false);
                else if (update.first == "buttonDOut22True")
                    SetDigitalOutput(21, true);
                else if (update.first == "buttonDOut22False")
                    SetDigitalOutput(21, false);
                else if (update.first == "buttonGSig2True")
                    SetGlobalSignal(1, true);
                else if (update.first == "buttonGSig2False")
                    SetGlobalSignal(1, false);
            }
        }
        // Handle text boxes
        else if (update.second->state().state_case() == robotcontrolapp::AppUIElement::AppUIState::kTextfieldState)
        {
            if (update.first == "textboxMotionProgramFile")
            {
                m_motionProgramFile = update.second->state().textfield_state().current_text();
            }
            else if (update.first == "textboxLogicProgramFile")
            {
                m_logicProgramFile = update.second->state().textfield_state().current_text();
            }
        }
        // Handle number boxes
        else if (update.second->state().state_case() == robotcontrolapp::AppUIElement::AppUIState::kNumberfieldState)
        {
            if (update.first == "numberboxMoveToJointA1")
            {
                m_moveToJointsA1Target = update.second->state().numberfield_state().current_number();
            }
            else if (update.first == "numberboxMoveToJointE1")
            {
                m_moveToJointsE1Target = update.second->state().numberfield_state().current_number();
            }
            else if (update.first == "numberboxMoveToJointSpeed")
            {
                m_moveToJointSpeed = update.second->state().numberfield_state().current_number();
                if (m_moveToJointSpeed < 0)
                {
                    m_moveToJointSpeed = 0;
                    SetNumber("numberboxMoveToJointSpeed", m_moveToJointSpeed);
                }
                else if (m_moveToJointSpeed > 100.0)
                {
                    m_moveToJointSpeed = 100.0;
                    SetNumber("numberboxMoveToJointSpeed", m_moveToJointSpeed);
                }
            }
            else if (update.first == "numberboxMoveToLinearX")
            {
                m_moveToCartXTarget = update.second->state().numberfield_state().current_number();
            }
            else if (update.first == "numberboxMoveToLinearE1")
            {
                m_moveToCartE1Target = update.second->state().numberfield_state().current_number();
            }
            else if (update.first == "numberboxMoveToLinearSpeed")
            {
                m_moveToCartSpeed = update.second->state().numberfield_state().current_number();
                if (m_moveToCartSpeed < 0)
                {
                    m_moveToCartSpeed = 0; 
                    SetNumber("numberboxMoveToLinearSpeed", m_moveToCartSpeed);
                }
            }
        }
    }
}

/**
 * @brief Translates a referencing state to a human readable string
 * @param state
 * @return
 */
std::string translateReferencingState(App::DataTypes::RobotState::ReferencingState state)
{
    switch (state)
    {
        default:
            return "n/a";
        case App::DataTypes::RobotState::ReferencingState::NOT_REFERENCED:
            return "not referenced";
        case App::DataTypes::RobotState::ReferencingState::IS_REFERENCED:
            return "referenced";
        case App::DataTypes::RobotState::ReferencingState::IS_REFERENCING:
            return "referencing...";
    }
}

/**
 * @brief Translates a program run state to a human readable string
 * @param runState
 * @return
 */
std::string translateProgramState(App::DataTypes::MotionState::RunState runState)
{
    switch (runState)
    {
        default:
            return "n/a";
        case App::DataTypes::MotionState::RunState::NOT_RUNNING:
            return "not running";
        case App::DataTypes::MotionState::RunState::RUNNING:
            return "running";
        case App::DataTypes::MotionState::RunState::PAUSED:
            return "paused";
    }
}

/**
 * @brief Updates the status UI
 */
void ControlApp::UpdateUI()
{
    // Section Initializing
    auto robotState = GetRobotState();
    QueueSetText("textHardwareState", robotState.hardwareState);
    QueueSetText("textReferencingStateAll", translateReferencingState(robotState.referencingState));
    QueueSetText("textReferencingStateA1", translateReferencingState(robotState.joints[0].referencingState));
    QueueSetText("textVelocityOverride", std::to_string((int)robotState.velocityOverride) + " %");

    // Section digital IO
    QueueSetText("textDIn22", robotState.digitalInputs.at(21) ? "ON" : "OFF"); // DIn22
    QueueSetText("textDOut22", robotState.digitalOutputs.at(21) ? "ON" : "OFF"); // DOut22
    QueueSetText("textGSig2", robotState.globalSignals.at(1) ? "ON" : "OFF"); // GSig2

    // Section Motion Program
    auto programState = GetMotionState();
    std::stringstream ssMotion;
    ssMotion << translateProgramState(programState.motionProgram.runState);
    if (programState.motionProgram.runState != App::DataTypes::MotionState::RunState::NOT_RUNNING)
    {
        ssMotion << ", in '" << programState.motionProgram.currentProgram << "' (" << (programState.motionProgram.currentProgramIndex + 1) << "/"
                 << programState.motionProgram.programCount << "), cmd " << (programState.motionProgram.currentCommandIndex + 1) << "/"
                 << programState.motionProgram.commandCount;
    }
    else {
        ssMotion << ", in '" << programState.motionProgram.currentProgram << "(not running)";
    }
    QueueSetText("textMotionProgramStatus", ssMotion.str());
    QueueSetText("textboxMotionProgramFile", programState.motionProgram.mainProgram);

    // Section Logic Program
    std::stringstream ssLogic;
    ssLogic << translateProgramState(programState.logicProgram.runState);
    if (programState.logicProgram.runState != App::DataTypes::MotionState::RunState::NOT_RUNNING)
    {
        ssLogic << ", in '" << programState.logicProgram.currentProgram << "' ("
                << (programState.logicProgram.currentProgramIndex + 1) << "/" << programState.logicProgram.programCount << "), cmd "
                << (programState.logicProgram.currentCommandIndex + 1) << "/" << programState.logicProgram.commandCount;
    }
    else
    {
        ssLogic << ", in '" << programState.logicProgram.currentProgram << "(not running)";
    }
    QueueSetText("textLogicProgramStatus", ssLogic.str());
    QueueSetText("textboxLogicProgramFile", programState.logicProgram.mainProgram);

    SendQueuedUIUpdates();
}

/**
 * @brief Increases the velocity override
 */
void ControlApp::ExampleFaster()
{
    SetVelocity(std::min(100.0f, GetVelocity() + 10.0f));
    SetText("textVelocityOverride", std::to_string((int)GetVelocity()) + " %");
}

/**
 * @brief Decreases the velocity override
 */
void ControlApp::ExampleSlower() {
    SetVelocity(std::max(0.0f, GetVelocity() - 10.0f));
    SetText("textVelocityOverride", std::to_string((int)GetVelocity()) + " %");
}

/**
 * @brief Example: Move to position by joint motion
 */
void ControlApp::ExampleMoveToJoint()
{
    // Only move A1 and E1 as given by the user, keep all other joints at their current position
    auto robotState = GetRobotState();
    MoveToJoint(m_moveToJointSpeed, 40, m_moveToJointsA1Target, robotState.joints[1].targetPosition, robotState.joints[2].targetPosition,
                robotState.joints[3].targetPosition, robotState.joints[4].targetPosition, robotState.joints[5].targetPosition, m_moveToJointsE1Target,
                robotState.joints[7].targetPosition, robotState.joints[8].targetPosition);
}

/**
 * @brief Example: Move to relative position by joint motion
 */
void ControlApp::ExampleMoveToJointRelative()
{
    MoveToJointRelative(m_moveToJointSpeed, 40, m_moveToJointsA1Target, 0, 0, 0, 0, 0, m_moveToJointsE1Target, 0, 0);
}

/**
 * @brief Example: Move to position by linear motion
 */
void ControlApp::ExampleMoveToCart()
{
    // Only move X and E1 as given by the user, keep all other axes at their current position
    auto robotState = GetRobotState();
    MoveToLinear(m_moveToCartSpeed, 40, m_moveToCartXTarget, robotState.tcp.GetY(), robotState.tcp.GetZ(), robotState.tcp.GetA(), robotState.tcp.GetB(),
                 robotState.tcp.GetC(), m_moveToCartE1Target, robotState.joints[7].targetPosition, robotState.joints[8].targetPosition);
}

/**
 * @brief Example: Move to relative position by linear motion
 */
void ControlApp::ExampleMoveToCartRelativeBase()
{
    MoveToLinearRelativeBase(m_moveToCartSpeed, 40, m_moveToCartXTarget, 0, 0, 0, 0, 0, m_moveToCartE1Target, 0, 0);
}

/**
 * @brief Example: Move to relative position by linear motion
 */
void ControlApp::ExampleMoveToCartRelativeTool()
{
    MoveToLinearRelativeTool(m_moveToCartSpeed, 40, m_moveToCartXTarget, 0, 0, 0, 0, 0, m_moveToCartE1Target, 0, 0);
}

/**
 * @brief Example: Upload sample program file from file
 */
void ControlApp::ExampleUploadSampleProgramFromFile() {
    std::cout << "Uploading sample file '" << m_sampleRemoteFileName << "' from file '" << m_sampleUploadFileName <<"'..." << std::endl;

    std::string errorMsg;
    if (UploadFile(m_sampleUploadFileName, m_sampleRemoteFileName, errorMsg))
    {
        std::cout << "Sample file '" << m_sampleUploadFileName << "' uploaded from file" << std::endl;
    }
    else
    {
        std::cout << "Could not upload sample file '" << m_sampleUploadFileName << "': " << errorMsg << std::endl;
    }
}

/**
 * @brief Example: Upload sample program file from memory
 */
void ControlApp::ExampleUploadSampleProgramFromMemory() {
    std::cout << "Uploading sample file '" << m_sampleRemoteFileName << "' from memory..." << std::endl;

    // Flash DOut21 in a 1s interval
    std::string content =
        "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
        "<Program>"
        "    <Output Nr=\"1\" Channel=\"DOut21\" State=\"True\" Descr=\"\" />"
        "    <Wait Nr=\"2\" Type=\"Time\" Seconds=\"1\" Descr=\"\" />"
        "    <Output Nr=\"3\" Channel=\"DOut21\" State=\"False\" Descr=\"\" />"
        "    <Wait Nr=\"4\" Type=\"Time\" Seconds=\"1\" Descr=\"\" />"
        "</Program>";

    std::string errorMsg;
    if (UploadFile((uint8_t*)content.data(), content.size(), m_sampleRemoteFileName, errorMsg))
    {
        std::cout << "Sample file '" << m_sampleRemoteFileName << "' uploaded from memory" << std::endl;
    }
    else
    {
        std::cout << "Could not upload sample file '" << m_sampleRemoteFileName <<"' from memory: " << errorMsg << std::endl;
    }
}

/**
 * @brief Example: Download sample program file to file
 */
void ControlApp::ExampleDownloadSampleProgramToFile()
{
    std::cout << "Downloading sample file '" << m_sampleRemoteFileName << "' to file '" << m_sampleDownloadFileName << "'..." << std::endl;

    std::string errorMsg;
    if (DownloadFile(m_sampleRemoteFileName, m_sampleDownloadFileName, errorMsg))
    {
        std::cout << "Sample file '" << m_sampleRemoteFileName << "' downloaded to file '" << m_sampleDownloadFileName << std::endl;
    }
    else
    {
        std::cerr << "Could not download sample file '" << m_sampleRemoteFileName << "' to file '" << m_sampleDownloadFileName << "': " << errorMsg
                  << std::endl;
    }
}

/**
 * @brief Example: Download sample program file to memory
 */
void ControlApp::ExampleDownloadSampleProgramToMemory()
{
    std::cout << "Downloading sample file '" << m_sampleRemoteFileName << "' to memory..." << std::endl;

    std::vector<uint8_t> data;
    std::string errorMsg;
    if (DownloadFile(m_sampleRemoteFileName, data, errorMsg))
    {
        std::cout << "Sample file '" << m_sampleRemoteFileName << "' downloaded to memory (" << data.size() << " bytes):" << std::endl;

        // convert binary data to string
        std::string dataStr(reinterpret_cast<char*>(data.data()), data.size());
        // print only the first bytes of the file
        if (dataStr.size() > 1024)
        {
            std::cout << dataStr.substr(0, 1024) << std::endl;
            std::cout << "..." << std::endl;
        }
        else
        {
            std::cout << dataStr << std::endl;
        }
    }
    else
    {
        std::cerr << "Could not download sample file '" << m_sampleRemoteFileName << "' to memory: " << errorMsg << std::endl;
    }
}

/**
 * @brief Example: List the files in the Programs directory
 */
void ControlApp::ExampleListPrograms()
{
    std::string directoryName = "Programs";
    AppClient::DirectoryContent files = ListFiles(directoryName);

    if (files.success)
    {
        std::cout << "Content of directory '" << directoryName << "' (" << files.entries.size() << " entries):" << std::endl;
        for (const auto& entry : files.entries)
        {
            switch (entry.type)
            {
                case robotcontrolapp::ListFilesResponse_DirectoryEntry_Type_File:
                    std::cout << "File:  ";
                    break;
                case robotcontrolapp::ListFilesResponse_DirectoryEntry_Type_Directory:
                    std::cout << "Dir:   ";
                    break;
                case robotcontrolapp::ListFilesResponse_DirectoryEntry_Type_Other:
                    std::cout << "Other: ";
                    break;
                default:
                    std::cout << "???:   ";
                    break;
            }
            std::cout << entry.name << std::endl;
        }
    }
    else
    {
        std::cerr << "Could not read directory '" << directoryName << "': " << files.errorMessage << std::endl;
    }
}

File: control_python/ControlApp.py
import sys
from AppClient import AppClient
from DataTypes.Matrix44 import Matrix44
from robotcontrolapp_pb2 import AppFunction, AppUIElement, ButtonState, CheckboxState, ListFilesResponse, ReferencingState, RunState
from google.protobuf.internal import containers as protobufContainers


class ControlApp(AppClient):
    """This is an example app implementation"""
    
    def __init__(self, appName: str, target: str):
        """Initializes the app. Pass the app name (as defined in rcapp.xml) and socket to connect to (default: "localhost:5000")"""
        self.__motionProgramFile = ""
        self.__logicProgramFile = ""
        self.__moveToJointsA1Target = 0
        self.__moveToJointsE1Target = 0
        self.__moveToCartXTarget = 0
        self.__moveToCartE1Target = 0
        self.__moveToJointSpeed = 0
        self.__moveToCartSpeed = 0
        self.__sampleRemoteFileName = ""
        self.__sampleUploadFileName = ""
        self.__sampleDownloadFileName = ""
        AppClient.__init__(self, appName, target)

    def _AppFunctionHandler(self, function: AppFunction):
        """Gets called on remote app function calls received from the robot control"""
        return

    def _UiUpdateHandler(self, updates: protobufContainers.RepeatedCompositeFieldContainer[AppUIElement]):
        """Gets called on remote UI update requests received from the robot control"""
        for update in updates:
            if update.state.HasField("button_state"):
                # Handle buttons
                isClicked = update.state.button_state == ButtonState.CLICKED
                if isClicked:
                    # Init
                    if update.element_name == "buttonReset":
                        self.ResetErrors()
                    elif update.element_name == "buttonEnable":
                        self.EnableMotors()
                    elif update.element_name == "buttonDisable":
                        self.DisableMotors()
                    elif update.element_name == "buttonReferenceAll":
                        self.ReferenceAllJoints()
                    elif update.element_name == "buttonReferenceA1":
                        self.ReferenceJoints(0)
                    elif update.element_name == "buttonReferenceProgram":
                        self.ReferenceAllJoints(True)

                    # Velocity override
                    elif update.element_name == "buttonFaster":
                        self.ExampleFaster()
                    elif update.element_name == "buttonSlower":
                        self.ExampleSlower()

                    # Programs
                    elif update.element_name == "buttonProgramStart":
                        self.StartMotionProgram()
                    elif update.element_name == "buttonProgramStop":
                        self.StopMotionProgram()
                    elif update.element_name == "buttonProgramPause":
                        self.PauseMotionProgram()
                    elif update.element_name == "buttonProgramSingle":
                        self.SetMotionProgramSingle()
                    elif update.element_name == "buttonProgramRepeat":
                        self.SetMotionProgramRepeat()
                    elif update.element_name == "buttonProgramStep":
                        self.SetMotionProgramStep()
                    elif update.element_name == "buttonMotionProgramLoad":
                        self.LoadMotionProgram(self.__motionProgramFile)
                    elif update.element_name == "buttonMotionProgramUnload":
                        self.UnloadMotionProgram()
                    elif update.element_name == "buttonLogicProgramLoad":
                        self.LoadLogicProgram(self.__logicProgramFile)
                    elif update.element_name == "buttonLogicProgramUnload":
                        self.UnloadLogicProgram()
                    
                    # Move To
                    elif update.element_name == "buttonMoveToStop":
                        self.MoveToStop()
                    elif update.element_name == "buttonMoveToJoint":
                        self.ExampleMoveToJoint()
                    elif update.element_name == "buttonMoveToJointRelative":
                        self.ExampleMoveToJointRelative()
                    elif update.element_name == "buttonMoveToCart":
                        self.ExampleMoveToCart()
                    elif update.element_name == "buttonMoveToCartBaseRelative":
                        self.ExampleMoveToCartRelativeBase()
                    elif update.element_name == "buttonMoveToCartToolRelative":
                        self.ExampleMoveToCartRelativeTool()
                    elif update.element_name == "buttonProgramUploadSampleFile":
                        self.ExampleUploadSampleProgramFromFile()
                    elif update.element_name == "buttonProgramUploadSampleMemory":
                        self.ExampleUploadSampleProgramFromMemory()
                    elif update.element_name == "buttonProgramDownloadSampleFile":
                        self.ExampleDownloadSampleProgramToFile()
                    elif update.element_name == "buttonProgramDownloadSampleMemory":
                        self.ExampleUploadSampleProgramFromMemory()
                    elif update.element_name == "buttonProgramList":
                        self.ExampleListPrograms()

                    # Digital IO
                    elif update.element_name == "buttonDIn22True":
                        self.SetDigitalInput(21, True)
                    elif update.element_name == "buttonDIn22False":
                        self.SetDigitalInput(21, False)
                    elif update.element_name == "buttonDOut22True":
                        self.SetDigitalOutput(21, True)
                    elif update.element_name == "buttonDOut22False":
                        self.SetDigitalOutput(21, False)
                    elif update.element_name == "buttonGSig2True":
                        self.SetGlobalSignal(1, True)
                    elif update.element_name == "buttonGSig2False":
                        self.SetGlobalSignal(1, False)
                        
            elif update.HasField("textfield_state"):
                # Handle text boxes
                if update.element_name == "textboxMotionProgramFile":
                    self.__motionProgramFile = update.state.textfield_state.current_text()
                elif update.element_name == "textboxLogicProgramFile":
                    self.__logicProgramFile = update.state.textfield_state.current_text()

            elif update.HasField("numberfield_state"):
                # Handle number boxes
                if update.element_name == "numberboxMoveToJointA1":
                    self.__moveToJointsA1Target = update.state.numberfield_state.current_number()
                elif update.element_name == "numberboxMoveToJointE1":
                    self.__moveToJointsE1Target = update.state.numberfield_state.current_number()
                elif update.element_name == "numberboxMoveToJointSpeed":
                    self.__moveToJointSpeed = update.state.numberfield_state.current_number()
                    if self.__moveToJointSpeed < 0:
                        self.__moveToJointSpeed = 0
                        self.SetNumber("numberboxMoveToJointSpeed", self.__moveToJointSpeed)
                    elif self.__moveToJointSpeed > 100:
                        self.__moveToJointSpeed = 100
                        self.SetNumber("numberboxMoveToJointSpeed", self.__moveToJointSpeed)
                elif update.element_name == "numberboxMoveToLinearX":
                    self.__moveToCartXTarget = update.state.numberfield_state.current_number()
                elif update.element_name == "numberboxMoveToLinearE1":
                    self.__moveToCartE1Target = update.state.numberfield_state.current_number()
                elif update.element_name == "numberboxMoveToLinearSpeed":
                    self.__moveToCartSpeed = update.state.numberfield_state.current_number()
                    if self.__moveToCartSpeed < 0:
                        self.__moveToCartSpeed = 0
                        self.SetNumber("numberboxMoveToLinearSpeed", self.__moveToCartSpeed)

    def translateReferencingState(state: ReferencingState) -> str:
        """Translates a referencing state to a human readable string"""
        if state == ReferencingState.NOT_REFERENCED:
            return "not referenced"
        elif state == ReferencingState.IS_REFERENCED:
            return "referenced"
        elif state == ReferencingState.IS_REFERENCING:
            return "referencing..."
        else:
            return "n/a"
      
    def translateProgramState(runState: RunState) -> str:
        """Translates a program run state to a human readable string"""
        if runState == RunState.NOT_RUNNING:
            return "not running"
        elif runState == RunState.RUNNING:
            return "running"
        elif runState == RunState.PAUSED:
            return "paused"
        else:
            return "n/a"

    def UpdateUI(self):
        """Updates the status UI"""
        
        # Section Initializing
        robotState = self.GetRobotState()
        self.QueueSetText("textHardwareState", robotState.hardwareState)
        self.QueueSetText("textReferencingStateAll", self.translateReferencingState(robotState.referencingState))
        self.QueueSetText("textReferencingStateA1", self.translateReferencingState(robotState.joints[0].referencingState))
        self.QueueSetText("textVelocityOverride", robotState.velocityOverride + " %")

        # Section digital IO
        if robotState.digitalInputs[21]: # DIn22
            self.QueueSetText("textDIn22", "ON")
        else:
            self.QueueSetText("textDIn22", "OFF")
        if robotState.digitalOutputs[21]: # DOut22
            self.QueueSetText("textDOut22", "ON" )
        else:
            self.QueueSetText("textDOut22", "OFF")
        if robotState.globalSignals[1]: # GSig2
            self.QueueSetText("textGSig2", "ON")
        else:
            self.QueueSetText("textGSig2", "OFF")

        # Section Motion Program
        programState = self.GetMotionState()
        motionStr = self.translateProgramState(programState.motionProgram.runState)
        if programState.motionProgram.runState != RunState.NOT_RUNNING:
            motionStr = motionStr + ", in '" + programState.motionProgram.currentProgram + "' (" +(programState.motionProgram.currentProgramIndex +1) + "/" + programState.motionProgram.programCount
            motionStr = motionStr + "), cmd " + (programState.motionProgram.currentCommandIndex + 1) + "/" + programState.motionProgram.commandCount
        else:
            motionStr = motionStr + ", in '" + programState.motionProgram.currentProgram + " (not running)"
        
        self.QueueSetText("textMotionProgramStatus", motionStr)
        self.QueueSetText("textboxMotionProgramFile", programState.motionProgram.mainProgram)

        # Section Logic Program
        logicStr = self.translateProgramState(programState.logicProgram.runState)
        if programState.logicProgram.runState != RunState.NOT_RUNNING:
            logicStr = logicStr + ", in '" + programState.logicProgram.currentProgram + "' (" +(programState.logicProgram.currentProgramIndex +1) + "/" + programState.motionProgram.programCount
            logicStr = logicStr + "), cmd " + (programState.logicProgram.currentCommandIndex + 1) + "/" + programState.logicProgram.commandCount
        else:
            logicStr = logicStr + ", in '" + programState.logicProgram.currentProgram + " (not running)"
        
        self.QueueSetText("textLogicProgramStatus", logicStr)
        self.QueueSetText("textboxLogicProgramFile", programState.logicProgram.mainProgram)

        # Send UI updates
        self.SendQueuedUIUpdates()

    def ExampleFaster(self):
        """Increases the velocity override"""
        self.SetVelocity(min(100, self.GetVelocity() + 10))
        self.SetText("textVelocityOverride", self.GetVelocity() + " %")
    
    def ExampleSlower(self):
        """Decreases the velocity override"""
        self.SetVelocity(max(0, self.GetVelocity() - 10))
        self.SetText("textVelocityOverride", self.GetVelocity() + " %")
    
    def ExampleMoveToJoint(self):
        """Example: Move to position by joint motion"""
        robotState = self.GetRobotState()
        self.MoveToJoint(self.__moveToJointSpeed, 40, self.__moveToJointsA1Target, robotState.joints[1].targetPosition, robotState.joints[2].targetPosition, robotState.joints[3].targetPosition, robotState.joints[4].targetPosition, robotState.joints[5].targetPosition, self.__moveToJointsE1Target, robotState.joints[7].targetPosition, robotState.joints[8].targetPosition)
    
    def ExampleMoveToJointRelative(self):
        """Example: Move to relative position by joint motion"""
        self.MoveToJointRelative(self.__moveToJointSpeed, 40, self.__moveToJointsA1Target, 0, 0, 0, 0, 0, self.__moveToJointsE1Target, 0 ,0)
    
    def ExampleMoveToCart(self):
        """Example: Move to position by linear motion"""
        robotState = self.GetRobotState()
        self.MoveToJoint(self.__moveToCartSpeed, 40, self.__moveToCartXTarget, robotState.tcp.GetY(), robotState.tcp.GetZ(),robotState.tcp.GetA(), robotState.tcp.GetB(), robotState.tcp.GetC(), self.__moveToCartE1Target, robotState.joints[7].targetPosition, robotState.joints[8].targetPosition)
    
    def ExampleMoveToCartRelativeBase(self):
        """Example: Move to relative position by linear motion"""
        self.MoveToLinearRelativeBase(self.__moveToCartSpeed, 40, self.__moveToCartXTarget, 0, 0, 0, 0, 0, self.__moveToCartE1Target, 0, 0)
    
    def ExampleMoveToCartRelativeTool(self):
        """Example: Move to relative position by linear motion"""
        self.MoveToLinearRelativeTool(self.__moveToCartSpeed, 40, self.__moveToCartXTarget, 0, 0, 0, 0, 0, self.__moveToCartE1Target, 0, 0)

    def ExampleUploadSampleProgramFromFile(self):
        """Example: Upload sample program file from file"""
        (success, errorMsg) = self.UploadFileFromFile(self.__sampleUploadFileName, self.__sampleRemoteFileName)
        if success:
            print("Sample file '" + self.__sampleRemoteFileName + "' uploaded from file '" + self.__sampleUploadFileName + "'...")
        else:
            print("Could not upload sample file '" + self.__sampleRemoteFileName + "': " << errorMsg)
    
    def ExampleUploadSampleProgramFromMemory(self):
        """Example: Upload sample program file from memory"""
        print("Uploading sample file '" + self.__sampleRemoteFileName + "' from memory...")
        
        # Flash DOut21 in a 1s interval
        content = "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
        "<Program>"
        "    <Output Nr=\"1\" Channel=\"DOut21\" State=\"True\" Descr=\"\" />"
        "    <Wait Nr=\"2\" Type=\"Time\" Seconds=\"1\" Descr=\"\" />"
        "    <Output Nr=\"3\" Channel=\"DOut21\" State=\"False\" Descr=\"\" />"
        "    <Wait Nr=\"4\" Type=\"Time\" Seconds=\"1\" Descr=\"\" />"
        "</Program>"

        (success, errorMsg) = self.UploadFileFromMemory(content, self.__sampleRemoteFileName)
        if success:
            print("Sample file '" + self.__sampleRemoteFileName + "' uploaded from memory")
        else:
            print("Could not upload sample file '" + self.__sampleRemoteFileName + "': " << errorMsg)
    
    def ExampleDownloadSampleProgramToFile(self):
        """Example: Download sample program file to file"""
        print("Downloading sample file '" + self.__sampleRemoteFileName + "' to file '" + self.__sampleDownloadFileName + "'..." )

        (success, errorMsg) = self.DownloadFileToFile(self.__sampleRemoteFileName, self.__sampleDownloadFileName)
        if success:
            print("Sample file '" + self.__sampleRemoteFileName + "' downloaded to file '" + self.__sampleDownloadFileName)
        else:
            print("Could not download sample file '" + self.__sampleRemoteFileName + "' to file '" + self.__sampleDownloadFileName + "': " + errorMsg)
    
    def ExampleDownloadSampleProgramToMemory(self):
        """Example: Download sample program file to memory"""
        print("Downloading sample file '" + self.__sampleRemoteFileName + "' to memory...")

        (success, errorMsg, data) = self.DownloadFileToMemory(self.__sampleRemoteFileName)
        if success:
            print("Sample file '" + self.__sampleRemoteFileName + "' downloaded to memory (" + data.count() + " bytes):")
            if(data.count > 1024):
                # Print only the first bytes of the file
                dataStr = str(data[:1024])
                print(dataStr)
            else:
                print(str(data))
        else:
            print("Could not download sample file '" + self.__sampleRemoteFileName + "' to memory: " + errorMsg)
    
    def ExampleListPrograms(self):
        """Example: List the files in the Programs directory"""
        directoryName = "Programs"
        files = self.ListFiles(directoryName)

        if(files.success):
            print("Content of directory '" + directoryName + "' (" + files.entries.count() + " entries):")

            for entry in files.entries:
                if entry.type == ListFilesResponse.DirectoryEntry.Type.File:
                    print("File: " + entry.name)
                elif entry.type == ListFilesResponse.DirectoryEntry.Type.Directory:
                    print("Directory: " + entry.name)
                elif entry.type == ListFilesResponse.DirectoryEntry.Type.Other:
                    print("Other: " + entry.name)
                else:
                    print("???: " + entry.name)
        else:
            print("Could not read directory '" + directoryName + "': " + files.errorMessage)